src/fp-paradigm/js/main.js

```diff
@@ -36,6 +41,57 @@ const render = (form) => {
 const compose = (...functions) => (value = null) =>
     functions.reduceRight((x, f) => f(x), value);
 
-const main = (inputs) => compose(render, renderForm, renderFormGroups)(inputs);
+const main = (inputs, submitHandler) => {
+    compose(render, renderForm, renderFormGroups)(inputs);
+
+    const form = getNodeBySelector('.form');
+    form.addEventListener('submit', submitHandler);
+};
+
+const getNodeBySelector = (selector) => document.querySelector(selector);
+const getValueFromInput = (selector) => document.querySelector(selector).value;
+const deleteTodo = (i) =>
+    dispatch({
+        type: 'DELETE_TODO',
+        id: i,
+    });
+const toggleTodo = (i) =>
+    dispatch({
+        type: 'TOGGLE_TODO',
+        id: i,
+    });
+
+const rerender = (todos) => {
+    const todosView = getNodeBySelector('.todos');
+    const taskInput = getNodeBySelector('.task-input');
+
+    taskInput.value = '';
+
+    const newTodosList = todos.reduce(
+        (nodes, todo, i) =>
+            (nodes += `
+            <div class="todo">
+                <div class="todo__id">${todo.id}</div>
+                <div class="todo__task">${todo.task}</div>
+                <div class="todo__completed">${todo.completed}</div>
+                <button onclick="deleteTodo(${i})")">Delete</button>
+                <button onclick="toggleTodo(${i})")">Completed</button>
+            </div>
+        `),
+        '',
+    );
+
+    todosView.innerHTML = newTodosList;
+};
```

Comment:
Yes, this is also imperative programming, and need to be fixed. And also this is **impure function**, which doesn't return anything. In fact, a function which doesn't return a thing, is called impure function or shortly **void functions**. Will be great, if we discuss about it more.